src/Corax/IndexSearcher.cs

```diff
@@ -187,6 +196,71 @@ public MultiTermMatch StartWithQuery(string field, string startWith)
+        }
+
+        public SortingMatch OrderByDescending<TInner>(in TInner set, int fieldId, MatchCompareFieldType entryFieldType = MatchCompareFieldType.Sequence, int take = -1)
+            where TInner : IQueryMatch
+        {
+            return OrderBy<TInner, DescendingMatchComparer>(in set, fieldId, MatchCompareFieldType.Sequence, take);
+        }
+
+        public SortingMatch OrderBy<TInner, TComparer>(in TInner set, int fieldId, MatchCompareFieldType entryFieldType = MatchCompareFieldType.Sequence, int take = -1)
+            where TInner : IQueryMatch
+            where TComparer : IMatchComparer
+        {
+            if (typeof(TComparer) == typeof(AscendingMatchComparer))
+            {
+                return Create(new SortingMatch<TInner, AscendingMatchComparer>(this, set, new AscendingMatchComparer(this, fieldId, entryFieldType), take));
+            }
+            else if (typeof(TComparer) == typeof(DescendingMatchComparer))
+            {
+                return Create(new SortingMatch<TInner, DescendingMatchComparer>(this, set, new DescendingMatchComparer(this, fieldId, entryFieldType), take));
+            }
+            else if (typeof(TComparer) == typeof(CustomMatchComparer))
+            {
+                throw new ArgumentException($"Custom comparers can only be created through the {nameof(OrderByCustomOrder)}");
+            }
+
+            throw new ArgumentException($"The comparer of type {typeof(TComparer).Name} is not supported. Isn't {nameof(OrderByCustomOrder)} the right call for it?");
+        }
+
+        public SortingMatch OrderBy<TInner, TComparer>(in TInner set, in TComparer comparer, int take = -1)
+            where TInner : IQueryMatch
+            where TComparer : struct, IMatchComparer
+        {
+            return Create(new SortingMatch<TInner, TComparer>(this, set, comparer, take));
+        }
+
+        public SortingMatch OrderByCustomOrder<TInner>(in TInner set, int fieldId, 
+                delegate*<IndexSearcher, int, long, long, int> compareByIdFunc,
+                delegate*<long, long, int> compareLongFunc,
+                delegate*<double, double, int> compareDoubleFunc,
+                delegate*<ReadOnlySpan<byte>, ReadOnlySpan<byte>, int> compareSequenceFunc,
+                MatchCompareFieldType entryFieldType = MatchCompareFieldType.Sequence, 
+                int take = -1)
+            where TInner : IQueryMatch     
+        {
+            // Federico: I don't even really know if we are going to find a use case for this. However, it was built for the purpose
+            //           of showing that it is possible to build any custom group of functions. Why would we want to do this instead
+            //           of just building a TComparer, I dont know. But for now the `CustomMatchComparer` can be built like this from
+            //           static functions. 
+            return Create(new SortingMatch<TInner, CustomMatchComparer>(
```

Comment:
Agreed, that it not something that we'll likely want and is candidate for removal.